File: Provenance/Services/PVAltKitService.m
//
//  PVAltKitService.m
//  Provenance
//
//  Created by Joseph Mattiello on 7/30/21.
//  Copyright © 2021 Provenance Emu. All rights reserved.
//

#if !TARGET_OS_OSX && !TARGET_OS_MACCATALYST && !TARGET_OS_TV
#import "PVAltKitService.h"
@import PVSupport;
@import AltKit;

@implementation PVAltKitService

+ (instancetype)sharedInstance {
    static dispatch_once_t onceToken;
    static PVAltKitService *sharedManager = nil;
    dispatch_once(&onceToken, ^{
        sharedManager = [PVAltKitService new];
    });
    return sharedManager;
}

- (void)start {
#if !TARGET_IPHONE_SIMULATOR
    [[ALTServerManager sharedManager] startDiscovering];

    [[ALTServerManager sharedManager]
        autoconnectWithCompletionHandler:^(ALTServerConnection *connection,
                                           NSError *error) {
          if (error) {
            return ELOG(@"Could not auto-connect to server. %@", error);
          }

          [connection enableUnsignedCodeExecutionWithCompletionHandler:^(
                          BOOL success, NSError *error) {
            if (success) {
              ILOG(@"Successfully enabled JIT compilation!");
              [[ALTServerManager sharedManager] stopDiscovering];
            } else {
              WLOG(@"Could not enable JIT compilation. %@", error);
            }

            [connection disconnect];
          }];
        }];
#endif
}

@end
#endif

File: Provenance/Services/PVAltKitService.swift
//
//  PVAltKitService.swift
//  Provenance
//
//  Created by Joseph Mattiello on 6/12/24.
//  Copyright © 2024 Provenance Emu. All rights reserved.
//

import Foundation
import PVSupport
import AltKit
import os

struct PVAltKitService {
    static let shared = PVAltKitService()

    private init() {}

    func start() {
#if !targetEnvironment(macCatalyst) && !targetEnvironment(tvOS)
        ALTServerManager.shared.startDiscovering()

        ALTServerManager.shared.autoconnect { connection, error in
            if let error = error as NSError? {
                os_log("Could not auto-connect to server: %@", log: .altKitService, type: .error, error)
            }

            DispatchQueue.global().async {
                connection?.enableUnsignedCodeExecution { success, _ in
                    if success {
                        os_log("Successfully enabled JIT compilation!", log: .altKitService, type: .info)
                        ALTServerManager.shared.stopDiscovering()
                    } else {
                        os_log("Could not enable JIT compilation.", log: .altKitService, type: .error)
                    }

                    DispatchQueue.main.async {
                        connection?.disconnect()
                    }
                }
            }
        }
        #else
        WLOG("PVAltKitService not supported on this platform.")
#endif
    }
}
